components/ulp/esp32ulp_mapgen.py

```diff
@@ -2,71 +2,103 @@
 # esp32ulp_mapgen utility converts a symbol list provided by nm into an export script
 # for the linker and a header file.
 #
-# SPDX-FileCopyrightText: 2016-2021 Espressif Systems (Shanghai) CO LTD
+# SPDX-FileCopyrightText: 2016-2024 Espressif Systems (Shanghai) CO LTD
 # SPDX-License-Identifier: Apache-2.0
-
-from __future__ import print_function
-
 import argparse
 import os
+import re
 import textwrap
 import typing
 
 UTIL = os.path.basename(__file__)
 
 
-def gen_ld_h_from_sym(f_sym: typing.TextIO, f_ld: typing.TextIO, f_h: typing.TextIO, base_addr: int) -> None:
+def name_mangling(name: str) -> str:
+    # Simple and dumb name mangling for namespaced name following GCC algorithm
+    ns, n = name.split('::')
+    return '_ZN{0}{1}{2}{3}E'.format(len(ns), ns, len(n), n)
+
+
+def gen_ld_h_from_sym(f_sym: typing.TextIO, f_ld: typing.TextIO, f_h: typing.TextIO, base_addr: int, prefix: str) -> None:
     f_ld.write(textwrap.dedent(
         f"""
         /* ULP variable definitions for the linker.
          * This file is generated automatically by {UTIL} utility.
          */
-        """
-    ))
-    f_h.write(textwrap.dedent(
-        f"""
-        /* ULP variable definitions for the compiler.
-         * This file is generated automatically by {UTIL} utility.
-         */
-        #pragma once
-        #ifdef __cplusplus
-        extern "C" {{
-        #endif
-        """
+        """  # noqa: E222
     ))
+    cpp_mode = False
+    var_prefix = prefix
+    namespace = ''
+    if '::' in prefix:
+        # C++ mode, let's avoid the extern "C" type and instead use namespace
+        f_h.write(textwrap.dedent(
+            f"""
+            /* ULP variable definitions for the compiler.
+             * This file is generated automatically by {UTIL} utility.
+             */
+            #pragma once
+            """  # noqa: E222
+        ))
+        namespace, var_prefix = prefix.split('::')
```

Comment:
This could fail for nested namespaces. Suggest using `prefix.rsplit('::', 1)`.

Reply:
I think nested namespace is not supported and I don't see high interest in trying to support them. It has implications for the name mangling algorithm used in the file to generate the linker's symbol name and also means a recursive algorithm to write `namespace X { namespace Y { ...`, because the "namespace X::Y" writing requires C++17. I'll add a test to return an error in case there are more than one "::", since it's not intended to be handled.

Reply:
Ok, make it more like this: if prefix is `A::B::C_`, the code will generate `A::B_C_` as prefix. This saves from implementing complex name mangling rules with nested namespace and avoid breaking the tool like it would have done before. I think it's ok to do our best here, I don't think anyone will try to use more than 1 level namespace anyway.